### ISSUE TITLE
Segmentation fault on large tensors

### ISSUE THREAD

**Issue by abmazitov** [lab-cosmo/sphericart#41]
While computing spherical harmonics on large tensors, one may face the segmentation fault with signal 11. The following example should reproduce the error.

```
import torch
import sphericart.torch as sphericart

if __name__ == '__main__':
    calculator = sphericart.SphericalHarmonics(l_max=12, normalized=True)
    vectors = 5.0 * torch.rand(100000000, 3)
    sp_harm = calculator.compute(vectors)
```
---

Naive memory profiling of this script gives the following feedback:

```
/usr/bin/time -v python test_sphericart.py 
Command terminated by signal 11
        Command being timed: "python test_sphericart.py"
        User time (seconds): 3.54
        System time (seconds): 1.09
        Percent of CPU this job got: 64%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:07.22
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 2578860
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 67800
        Voluntary context switches: 8953
        Involuntary context switches: 97625
        Swaps: 0
        File system inputs: 0
        File system outputs: 0
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

**Comment by ceriottm** [comment]
Where are you running this? Looks like the return array would be roughly 400GB

**Comment by abmazitov** [comment]
This is basically an issue I face while doing the training with lbfgs on a large dataset with 10K structures and more. It requires all the data to be in a single batch, and that results in a huge xyz tensor.

**Comment by ceriottm** [comment]
What I mean is: if it's segfaulting because you try to allocate a vector that is larger than the RAM, then there's not much that can be done code-wise (other than perhaps give a better error message).

**Comment by abmazitov** [comment]
I tried this one on a jed bigmem node l, there should be enough memory to handle such a large tensor. 

Moreover, the error I'm getting is not OOM error, it is rather the code is allocating memory in a wrong way, which may be caused for instance by int32 array size variable overflow.

**Comment by ceriottm** [comment]
I see. So yes, might be that we are reaching the limits of int32. Refactoring everything with int64 will be quite a pain though. Let's discuss this IRL with also @frostedoyster and @Luthaf

**Comment by Luthaf** [comment]
The https://github.com/lab-cosmo/sphericart/tree/64-bit-int branch tries to replace all `int` with `int64_t` to check if it helps. This was easier to do than anticipated thanks to the public API/Python API already using the right `size_t` type.

> Moreover, the error I'm getting is not OOM error, it is rather the code is allocating memory in a wrong way, which may be caused for instance by int32 array size variable overflow.

Unfortunately, dues to [overcommit](https://en.wikipedia.org/wiki/Memory_overcommitment) on Linux, OOM might manifest as bus error or segfault as well.

**Comment by ceriottm** [comment]
@abmazitov did you check if the 64-bit-int branch fixes this problem?

**Comment by abmazitov** [comment]
Yes I did, it doesn't help.

**Comment by frostedoyster** [comment]
`
import numpy as np
import sphericart as sphericart

if __name__ == '__main__':
    n_samples = 15000000
    l_max = 10
    calculator = sphericart.SphericalHarmonics(l_max=l_max, normalized=True)
    if n_samples*(l_max+1)**2 / 10**9 < 2.147483647:
        print("Won't break due to int32")
    else:
        print("Will break due to int32")
    vectors = np.random.rand(n_samples, 3)
    sp_harm = calculator.compute(vectors)
`

**Comment by Luthaf** [comment]
We should run this code inside gdb/lldb after compiling with debug symbols, it should give us the exact line which is segfaulting.

With lldb (gdb should be similar, but I don't use it as much), this would mean adding `-DCMAKE_BUILD_TYPE=Debug` to https://github.com/lab-cosmo/sphericart/blob/70f139ae22f49b7d8a46d0bbbc735be6650ff3d9/setup.py#L39, recompiling, and then running `lldb python -- script-which-segfaults.py`, and once inside lldb using the `run` command.

**Comment by Luthaf** [comment]
Alternatively, we could re-compile with [ASAN](https://clang.llvm.org/docs/AddressSanitizer.html) and run the same code directly, ASAN should be able to track this kind of issues.

**Comment by frostedoyster** [comment]
I didn't manage to run it from Python, but here is a C++ program and relative gdb output

```
#include "sphericart.hpp"
#include <iostream>

int main() {
    int l_max = 10;
    int n_samples = 100000000;
    auto xyz = std::vector<double>(n_samples * 3, 0.0);
    std::vector<double> sph;
    auto calculator = sphericart::SphericalHarmonics<double>(l_max);
    calculator.compute(xyz, sph); // no gradients

    return 0;
}
```

```
Thread 22 "break_sphericart" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fe8d3d5e700 (LWP 3705824)]
0x0000000000410486 in generic_sph_sample<double, false, false, false, 6> (xyz_i=0x7fff914b2748, sph_i=0x7fe770adb338, dsph_i=0x0, ddsph_i=0x0, l_max=10, size_y=121, pylm=0x66bed0, pqlm=0x66c0e0, c=0x6744f0, s=0x674700, twomz=0x674910)
    at /home/bigi/sphericart/sphericart/include/templates.hpp:726
726         HARDCODED_SPH_MACRO(HARDCODED_LMAX, x, y, z, x2, y2, z2, sph_i, DUMMY_SPH_IDX);
```

**Comment by ceriottm** [comment]
PB: this needs to be run on a large memory node, so debugging is a bit of a pain.

**Comment by frostedoyster** [comment]
It's pretty clear that (and why) this was the issue:
```
sph_i = sph + i_sample * size_y;
```
where `sph_i` is a pointer and `i_sample`, `size_y` are ints. Changing `i_sample` to size_t fixes everything, while keeping the `l`-related variables as ints will hopefully let us retain all the computational advantages of ints.
See #67.
